### PR TITLE
Remove core packages that are build dependencies

### PR DESCRIPTION
**Pull request by uberhacker**
This PR removes a lot of extra build only dependencies that are not needed to execute binaries.

### CODE REVIEW

File: crew [hunk 1]
```diff
@@ -817,7 +817,12 @@ def install
     meta = download
     target_dir = unpack meta
     if meta[:source] == true
-      abort "You don't have a working C compiler. Run `crew install buildessential` to get one and try again.".lightred unless system('gcc --version')
+      abort "You don't have a working C compiler. Execute `crew install buildessential` and try again.".lightred unless system('gcc --version')
+      abort "You don't have a working Clang compiler. Execute `crew install buildessential` and try again.".lightred unless system('clang --version')
+      abort "You don't have a working CMake compiler. Execute `crew install buildessential` and try again.".lightred unless system('cmake --version')
```
Comment (saltedcoffii) [inline crew:822]: CMake isn't a compiler it's a build system

**Review by saltedcoffii — COMMENTED** [review]
I'm worried that suddenly a lot of packages aren't going to build properly. Might not hurt to go though and add depends_on => :build lines to packages

File: crew (hunk 1 above)
Comment (uberhacker) [reply crew:822]: According to Wikipedia, "CMake is not a build system but rather it generates another system's build files."  The actual cmake command does compile source code.

Comment (uberhacker) [reply crew:822]: The lines above are there to remove the need to add `depends_on 'something' => :build` to all the packages.  In other words, you won't be able to compile anything until buildessential is installed.

Comment (saltedcoffii) [reply crew:822]: Alright. According to the cmake website, it's a metabuild system

File: crew [hunk 2]
```diff
@@ -817,7 +817,12 @@ def install
     meta = download
     target_dir = unpack meta
     if meta[:source] == true
-      abort "You don't have a working C compiler. Run `crew install buildessential` to get one and try again.".lightred unless system('gcc --version')
+      abort "You don't have a working C compiler. Execute `crew install buildessential` and try again.".lightred unless system('gcc --version')
+      abort "You don't have a working Clang compiler. Execute `crew install buildessential` and try again.".lightred unless system('clang --version')
```
Comment (supechicken) [inline crew:821]: Better to add them to the dependency array automatically on building (if not installed)?

File: install.sh [hunk 3]
```diff
@@ -44,6 +46,9 @@ case "${ARCH}" in
   exit 1;;
 esac
 
+# install a base set of development packages for compiling and building software
+echo -n "Install development tools? (not needed for most users) [y/N]: "; read -n1 devtools
```
Comment (supechicken) [inline install.sh:50]: Merge them in a command?
```shell
read -n1 -p "${BLUE}Install development tools? (not needed for most users) [y/N]: ${RESET}" devtools
```

**Comment by satmandu** [comment]
two things: 

Here's my current WIP `install.sh` script I'm using to create my build containers, which uses package groups based upon early dependencies:

```
#!/bin/bash

# exit on fail
set -e

# export variables
set -a

#chromebrew directories
OWNER="${OWNER:-skycocker}"
REPO="${REPO:-chromebrew}"
BRANCH="${BRANCH:-master}"
URL="https://raw.githubusercontent.com/${OWNER}/${REPO}/${BRANCH}"
CREW_PREFIX="${CREW_PREFIX:-/usr/local}"
CREW_LIB_PATH="${CREW_PREFIX}/lib/crew/"
CREW_CONFIG_PATH="${CREW_PREFIX}/etc/crew/"
CREW_BREW_DIR="${CREW_PREFIX}/tmp/crew/"
CREW_DEST_DIR="${CREW_BREW_DIR}dest"
CREW_PACKAGES_PATH="${CREW_LIB_PATH}packages"
CREW_CACHE_DIR="${CREW_CACHE_DIR:-/home/chronos/user/.cache/crewcache}"
CURL="${CURL:-curl}"
PATH="${CREW_PREFIX}/bin:${CREW_PREFIX}/sbin:$PATH"

RUBY_PACKAGES="ncurses zlibpkg glibc gmp gcc10 libdb libffi libyaml \
ruby openssl"
CREW_PACKAGES="automake python3"
PARALLEL_PACKAGES="moreutils"
GIT_PACKAGES="git expat gdbm"
LLVM_STRIP_PACKAGES="llvm libedit gdbm libxml2 icu4c"
MAN_PACKAGES="gdbm"
EARLY_PACKAGES="$OLD_EARLY_PACKAGES $RUBY_PACKAGES $CREW_PACKAGES $PARALLEL_PACKAGES \
$GIT_PACKAGES $LLVM_STRIP_PACKAGES $MAN_PACKAGES"

OLD_EARLY_PACKAGES="gcc10 llvm brotli c_ares libcyrussasl libiconv libidn2 \
libmetalink libnghttp2 libpsl libssh2 libtirpc libunistring lz4 \
openldap rtmpdump zstd ncurses ca_certificates libyaml ruby libffi \
openssl nettle krb5 p11kit libtasn1 gnutls curl git icu4c libedit gdbm \
libpipeline python3 util_linux moreutils libseccomp coreutils automake \
readline glibc"

ARCH="$(uname -m)"

GREEN='\e[1;32m';
RED='\e[1;31m';
BLUE='\e[1;34m';
YELLOW='\e[1;33m';
RESET='\e[0m'
echo -e "${YELLOW}Starting Chromebrew install${RESET}"

if [ "${EUID}" == "0" ]; then
  echo -e "${RED}Chromebrew should not be installed or run as root.${RESET}"
  exit 1;
fi

case "${ARCH}" in
"i686"|"x86_64"|"armv7l"|"aarch64")
  LIB_SUFFIX=
  [ "${ARCH}" == "x86_64" ] && LIB_SUFFIX='64'
  ;;
*)
  echo -e "${RED}Your device is not supported by Chromebrew yet :/${RESET}"
  exit 1;;
esac

CREW_LIB_PREFIX="${CREW_PREFIX}/lib${LIB_SUFFIX}"
# Start with standard ChromeOS LD_LIBRARY_PATH
export LD_LIBRARY_PATH="/usr/lib${LIB_SUFFIX}:/lib${LIB_SUFFIX}"

# This will allow things to work without sudo
crew_folders="bin cache doc docbook etc include lib lib$LIB_SUFFIX libexec man sbin share tmp var"
for folder in $crew_folders
do
  if [ -d "${CREW_PREFIX}"/"${folder}" ]; then 
    # id -u gives "Illegal instruction" on zork
    sudo chown -R 1000:1000 "${CREW_PREFIX}"/"${folder}"
    #sudo chown -R "$(id -u)":"$(id -g)" "${CREW_PREFIX}"/"${folder}"
  fi
done
sudo chown 1000:1000 "${CREW_PREFIX}"
#sudo chown "$(id -u)":"$(id -g)" "${CREW_PREFIX}"
# Delete 'var' symlink on Cloudready platform
if ! grep -q neverware /etc/lsb-release; then
  [ -L /usr/local/var ] && sudo rm -f /usr/local/var
  [ -L /usr/local/local ] && sudo rm -f /usr/local/local
fi

# prepare directories
for dir in "${CREW_CONFIG_PATH}/meta" "${CREW_DEST_DIR}" "${CREW_PACKAGES_PATH}"; do
  if [ ! -d "${dir}" ]; then
    mkdir -p "${dir}"
  fi
done


case "${ARCH}" in
"i686")
  CURL_URL=https://github.com/moparisthebest/static-curl/releases/download/v7.75.0/curl-i386
  ;;
"x86_64")
  CURL_URL=https://github.com/moparisthebest/static-curl/releases/download/v7.75.0/curl-amd64
  ;;
"aarch64")
  CURL_URL=https://github.com/moparisthebest/static-curl/releases/download/v7.75.0/curl-aarch64
  ;;
"armv7l"|"armv8l")
  CURL_URL=https://github.com/moparisthebest/static-curl/releases/download/v7.75.0/curl-armv7
  ;;
*)
  echo -e "${RED}Your device is not supported by Chromebrew yet :/${RESET}"
  exit 1;;
esac

# Get static curl and use it for rest of install.
mkdir -p "${CREW_PREFIX}/bin"
echo -e "${BLUE}Downloading static curl for install...${RESET}"
LD_LIBRARY_PATH=/usr/lib${LIB_SUFFIX} /usr/bin/curl -# -Ls "$CURL_URL" \
 -o "${CREW_PREFIX}/bin/curl-static-${ARCH}" && \
chmod +x "${CREW_PREFIX}/bin/curl-static-${ARCH}" && \
export CURL="${CREW_PREFIX}/bin/curl-static-${ARCH}"
# export CURL="/usr/bin/curl"


# prepare url and sha256
urls=()
temp_url=
sha256s=()
temp_sha256=
k=0

case "${ARCH}" in
"armv7l"|"aarch64")
  if ! type "xz" > /dev/null; then
    temp_url='https://github.com/snailium/chrome-cross/releases/download/v1.8.1/xz-5.2.3-chromeos-armv7l.tar.gz'
    temp_sha256='4dc9f086ee7613ab0145ec0ed5ac804c80c620c92f515cb62bae8d3c508cbfe7'
    urls[k]="$temp_url"
    sha256s[k]="$temp_sha256"
    k=$((k+1))
  fi
  ;;
esac

for package in $EARLY_PACKAGES; do
  echo -e "${BLUE}Downloading ${package} package file...${RESET}"
  pkgfile="$($CURL -# -Lsf "${URL}"/packages/"$package".rb)"
  temp_url="$(echo "$pkgfile" | grep -m 3 "$ARCH": | head -n 1 | cut -d\' -f2 | tr -d \' | tr -d \" | sed 's/,//g')"
  temp_sha256="$(echo "$pkgfile" | grep -m 3 "$ARCH": | tail -n 1 | cut -d\' -f2 | tr -d \' | tr -d \" | sed 's/,//g')"
  urls[k]="$temp_url"
  sha256s[k]="$temp_sha256"
  k=$((k+1))
done

# functions to maintain packages
function download_check () {
    cd "${CREW_BREW_DIR}"

    if [ -f ${CREW_CACHE_DIR}/"${3}" ] && (echo "${4}" ${CREW_CACHE_DIR}/"${3}" | sha256sum -c -); then
      echo -e "${GREEN}Found ${1} in cache...${RESET}"
      mv "${CREW_CACHE_DIR}"/"${3}" "${CREW_BREW_DIR}"/"${3}"
    else
      #download
      echo -e "${BLUE}Downloading ${1}...${RESET}"
      $CURL '-#' -C - -L --ssl "${2}" -o "${3}"
      #verify
      echo -e "${BLUE}Verifying ${1}...${RESET}"
      echo -e "${GREEN}$(echo "${4}" "${3}" | sha256sum -c -)${RESET}"
      case "${?}" in
      0) ;;
      *)
        echo -e "${RED}Verification failed, something may be wrong with the download.${RESET}"
        exit 1;;
      esac
    fi
}

function extract_install () {
    # Start with a clean slate
    rm -rf "${CREW_DEST_DIR}"
    mkdir "${CREW_DEST_DIR}"
    cd "${CREW_DEST_DIR}"

    #extract and install
    echo "Extracting ${1} (this may take a while)..."
    tar xpf ../"${2}"
    echo "Installing ${1} (this may take a while)..."
    tar cpf - ./*/* | (cd /; tar xp --keep-directory-symlink -f -)
    mv ./dlist "${CREW_CONFIG_PATH}/meta/${1}.directorylist"
    mv ./filelist "${CREW_CONFIG_PATH}/meta/${1}.filelist"
    sudo ldconfig > /dev/null 2> /dev/null || true
}

function update_device_json () {
  cd "${CREW_CONFIG_PATH}"

  if grep '"name": "'"${1}"'"' device.json > /dev/null; then
    echo "Updating version number of ${1} in device.json..."
    sed -i device.json -e '/"name": "'"${1}"'"/N;//s/"version": ".*"/"version": "'"${2}"'"/'
  elif grep '^    }$' device.json > /dev/null; then
    echo "Adding new information on ${1} to device.json..."
    sed -i device.json -e '/^    }$/s/$/,\
    {\
      "name": "'"${1}"'",\
      "version": "'"${2}"'"\
    }/'
  else
    echo "Adding new information on ${1} to device.json..."
    sed -i device.json -e '/^  "installed_packages": \[$/s/$/\
    {\
      "name": "'"${1}"'",\
      "version": "'"${2}"'"\
    }/'
  fi
}

function get_deps () {
    for i in $(cat /tmp/sorted.txt)
  do
    echo "$i buildessential" >> /tmp/depraw.txt
    deps=$(cd packages ; grep "^[[:blank:]] depends_on '" "$i".rb | sed '/:build/d' | awk '{print $2}' | sed "s;';;g")
    for j in $deps
    do
      echo "$j $i" >> /tmp/depraw.txt
    done
  done
  awk '!x[$0]++' /tmp/depraw.txt > /tmp/deps.txt
  tsort /tmp/deps.txt | sed '/buildessential/d' | tac | awk '!x[$0]++' > /tmp/sorted.txt
}

# create the device.json file if it doesn't exist
cd "${CREW_CONFIG_PATH}"
if [ ! -f device.json ]; then
  echo -e "${YELLOW}Creating new device.json...${RESET}"
  echo '{' > device.json
  echo '  "architecture": "'"${ARCH}"'",' >> device.json
  echo '  "installed_packages": [' >> device.json
  echo '  ]' >> device.json
  echo '}' >> device.json
fi

for i in $(seq 0 $((${#urls[@]} - 1))); do
  url="${urls["${i}"]}"
  sha256="${sha256s["${i}"]}"
  tarfile="$(basename "${url}")"
  name="${tarfile%%-*}"   # extract string before first '-'
  rest="${tarfile#*-}"    # extract string after first '-'
  version="$(echo "${rest}" | sed -e 's/-chromeos.*$//')"
                        # extract string between first '-' and "-chromeos"

  download_check "${name}" "${url}" "${tarfile}" "${sha256}"
  extract_install "${name}" "${tarfile}"
  update_device_json "${name}" "${version}"

done

## workaround https://github.com/skycocker/chromebrew/issues/3305
sudo ldconfig > /dev/null 2> /dev/null || true
echo "$CREW_LIB_PREFIX" > "$CREW_PREFIX"/etc/ld.so.conf && \
ldconfig "$CREW_PREFIX"/etc/ld.so.conf -C "$CREW_PREFIX"/etc/ld.so.cache
export LD_LIBRARY_PATH=$CREW_LIB_PREFIX


# create symlink to 'crew' in ${CREW_PREFIX}/bin/
rm -f "${CREW_PREFIX}/bin/crew"
ln -s "../lib/crew/crew" "${CREW_PREFIX}/bin/"

# prepare sparse checkout .rb packages directory and do it
cd "${CREW_LIB_PATH}"
rm -rf .git
git init
git remote add -f origin "https://github.com/${OWNER}/${REPO}.git"
git config core.sparsecheckout true
echo packages >> .git/info/sparse-checkout
echo lib >> .git/info/sparse-checkout
echo crew >> .git/info/sparse-checkout
git fetch origin "${BRANCH}"
git reset --hard origin/"${BRANCH}"
[[ "$CREW_PREFIX" == "/usr/local" ]] && unset CREW_PREFIX
crew update

crew postinstall $EARLY_PACKAGES

# Use parallel to download essential packages, then install.
corefilepkgs=$(curl -Ls https://github.com/skycocker/chromebrew/raw/master/tools/core_packages.txt )
buildessentialpkgs=$(grep "^[[:blank:]] depends_on" packages/buildessential.rb | awk '{print $2}' | sed "s;';;g" )
echo "$EARLY_PACKAGES $corefilepkgs $buildessentialpkgs less most" | tr '\n' ' ' > /tmp/sorted.txt
for i in {1..7}
do
  echo "Dependency Search Pass $i"
  [[ -f /tmp/depraw.txt ]] && rm /tmp/depraw.txt
  get_deps
done
cat /tmp/sorted.txt
# Don't re-download EARLY_PACKAGES
for i in $EARLY_PACKAGES
do
  sed -i "/$i/d" /tmp/sorted.txt
done
sorted=$(cat /tmp/sorted.txt)
rm /tmp/{sorted.txt,deps.txt,depraw.txt}
echo "corefiles & buildessentials sorted in 1 level dependency order: $sorted"
parallel -j 4 CURL="$CURL" crew -d download -- "$sorted" || true
for i in  libjpeg gdk_pixbuf $sorted
do
yes |  CURL=$CURL crew install $i || true
done

echo
if [ -n "${CREW_PREFIX}" ]; then
  echo -e "${YELLOW}
Since you have installed Chromebrew in a directory other than '/usr/local',
you need to run these commands to complete your installation:
${RESET}"

  echo -e "${BLUE}
echo 'export CREW_PREFIX=${CREW_PREFIX}' >> ~/.bashrc
echo 'export PATH=\"\${CREW_PREFIX}/bin:\${CREW_PREFIX}/sbin:\${PATH}\"' >> ~/.bashrc
echo 'export LD_LIBRARY_PATH=${CREW_PREFIX}/lib${LIB_SUFFIX}' >> ~/.bashrc
source ~/.bashrc
${RESET}"
fi
echo -e "${BLUE}
To set the default PAGER environment variable to use less:
echo \"export PAGER='less'\" >> ~/.bashrc && . ~/.bashrc

Alternatively, you could use most.  Why settle for less, right?
echo \"export PAGER='most'\" >> ~/.bashrc && . ~/.bashrc

Below are some text editor suggestions.

To install 'nano', execute:
crew install nano

Or, to get an updated version of 'vim', execute:
crew install vim

You may wish to set the EDITOR environment variable for an editor default.

For example, to set 'nano' as the default editor, execute:
echo \"export EDITOR='nano'\" >> ~/.bashrc && . ~/.bashrc

To set 'vim' as the default editor, execute:
echo \"export EDITOR='vim'\" >> ~/.bashrc && . ~/.bashrc
${RESET}"
echo -e "${GREEN}Chromebrew installed successfully and package lists updated.${RESET}"
```
Here's the separate `getdeps` script I have been using, which you could modify to use a different core packages list... **but that will have to be run on every single package** to determine which core deps aren't needed. (This is why I think it is useful to have all runtime deps, including `core` deps, always listed. Changing your core packages list then doesn't require an en masse reevalauation of which packages now have missing deps, and if a package has a dep listed which is in core, it is harmless because it doesn't have to get reinstalled anyways.)
```
#!/bin/env bash
# set -x
pkg="${1}"
CREW_LIB_PREFIX=$(crew const | grep CREW_LIB_PREFIX | awk -F = '{print $2}')

# Which packages have a needed library in CREW_LIB_PREFIX
# This is a subset of what crew whatprovides gives
whatprovidesfxn() {
  pkgdepslcl="${1}"
  filelcl=$(grep "$pkgdepslcl" /usr/local/etc/crew/meta/*.filelist | grep $CREW_LIB_PREFIX)
  packagelcl=$(echo "$filelcl" | \
  sed 's/.filelist.*//g' | sed 's:.*/::' | awk '!x[$0]++' | sed s/://g)
  echo $packagelcl
}

# What files does a package provide
crewfilesfxn() {
 pkgname="${1}"
 files=$(cat /usr/local/etc/crew/meta/${pkgname}.filelist)
 echo $files
}

pkgfiles=$(crewfilesfxn "${pkg}")
# Use readelf to determine library dependencies. (This doesn't almost 
# run a program like ldd would.
pkgdeps=$(unset lines ; for i in $pkgfiles ; \
do lines+=$(readelf -d "$i" 2>/dev/null | grep NEEDED | awk '{print $5}' \
| sed 's/\[//g' | sed 's/\]//g') ; done ; echo "$lines" | tr " " "\n" \
| awk '!x[$0]++')


pkgdeps=$(unset lines; for j in $pkgdeps ; \
do lines+=$(echo -e "\n" ; whatprovidesfxn $j); done ; \
echo "$lines" | tr " " "\n" | awk '!x[$0]++')

# Remove original package from list.
pkgdeps=$(echo $pkgdeps | tr " " "\n" | sed "/${1}/d")
pkgdepsnocore=$pkgdeps
for i in $(curl -Ls https://github.com/skycocker/chromebrew/raw/master/tools/core_packages.txt) freetype_sub
do
  file=$(basename $i)
  pkgdepsnocore=$(echo $pkgdepsnocore | tr " " "\n" | sed "/$file/d")
done
pkgdepsnocore=$(echo $pkgdepsnocore | tr " " "\n"  | sort -u )

# If no non-core packages are dependencies, then just mention the core 
# dependencies
if [ -z "${pkgdepsnocore}" ]; then
pkgdepsdisplay=$pkgdeps
else
pkgdepsdisplay=$pkgdepsnocore
fi

for i in $pkgdepsdisplay
do
echo "  depends_on '$i'"
done
```

File: crew (hunk 2 above)
Comment (uberhacker) [reply crew:821]: Again, if you try to build and buildessential is not installed, you won't get that far.

File: install.sh (hunk 3 above)
Comment (uberhacker) [reply install.sh:50]: Good suggestion.

**Comment by uberhacker** [comment]
Good thoughts @satmandu.  I was thinking it would be very cool if there was a `crew getdeps <package>` command that could recursively search the dependency tree and report all.  This could also be incorporated into the currently lacking dependency algorithm in `crew` that sorely needs refactored.

File: crew (hunk 2 above)
Comment (uberhacker) [reply crew:821]: I see what you are saying but that would require a ton of package updates.

**Comment by satmandu** [comment]
@uberhacker that would be useful, but walking the dependency tree is different from walking actual library dependencies too, and we would want to separate build does from that list.

Putting library deps into a metadata file during build would also help...

Also note that the `install.sh` above uses the `tsort` app...

File: crew (hunk 2 above)
Comment (supechicken) [reply crew]: We can add them to `@deps` array in `expand_dependencies` if build is needed

Comment (uberhacker) [reply crew]: Oh, I get you now.  I didn't understand you meant to modify crew.  That does make sense.

Comment (uberhacker) [reply crew]: @supechicken: Thank you for reviewing.  I was afraid this was going to get lost in the PR queue like the others.  I made your suggested change.

**Review by satmandu — APPROVED** [review]

**Pull request merged**